<a href="https://colab.research.google.com/github/thomreid11/CMP7005_PRES1_Repository/blob/main/CMP7005_PRES1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Thomas Reid - CMP7005 - PRES1**

Exploratory Data Analysis (EDA) on India Air Quality Dataset

This EDA will investigate the air quality of 26 different Indian cities from 2017 to 2020. It will address health and environmental concerns linked to poor air quality in one of the worlds most polluted countries. This analysis will generate insghts into the different factors affecting air quality and the way in which different cities are impacted by pollution

#### **Importing the required libraries for the EDA**

In [44]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
import seaborn as sns

#### **Mounting Google Drive to Google Collab and importing dataset**

In [45]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [46]:
drive_path = '/content/drive/MyDrive/India Air Quality Dataset'

#### **Merging all csv files in folder into singular csv file using dataframes**

This process takes all 26 csv files from the 'drive_path' and reads them into a identical dataframes, using Pandas. Once processed, each file is concatenated into one dataframe. A new csv files is produced with all datasets.

In [47]:
dataframes = []
# For the files in the foler drive_path, if it ends in csv, then the file is appended into the dataframe
for filename in os.listdir(drive_path):
    if filename.endswith('.csv'):
        file_path = os.path.join(drive_path, filename)
        df = pd.read_csv(file_path)
        dataframes.append(df)
# Concatenating dataframes
df = pd.concat(dataframes,ignore_index=True)

In [48]:
# Dataframe to csv file
df_to_csv = df.to_csv('air_quality_data.csv')

#### **General Data Insights**

This section explores fundamental data understanding through examining basic attributes of the dataset such as data types and descriptive statistics.

**Key**

City - The name of the location where the air quality measurements were recorded

Date - The specific day on which the data was collected

PM2.5 - Concentration of fine particulate matter smaller than 2.5 micrometers

PM10 - Concentration of particulate matter smaller than 10 micrometers. From dust, pollen

NO - Nitric oxide concentration. From combustion processes

NO2 - Nitrogen dioxide concentration. From vehicles and industrial activities

NOx - Combined concentration of nitrogen oxides (NO + NO2). Leads to smog and respiratory issues

NH3 - Ammonia concentration. From agricultural and industrial emissions

CO - Carbon monoxide concentration. From incomplete combustion

SO2 - Sulfur dioxide concentration. From coal burning

O3 - Ozone concentration. Formed in the atmosphere, high levels indicate poor air quality

Benzene - Concentration of benzene. Carcinogenic

Toluene - Concentration of toluene. Solvent from industrial emissions

Xylene - Concentration of xylene. From vehicle exhaust

AQI - Air Quality Index value calculated from pollutant levels, representing overall air quality.

AQI_Bucket - Category describing the air quality level (e.g., "Good", "Satisfactory", "Moderate", "Poor", etc.) based on the AQI value.




In [49]:
# First 5 rows of dataframe
df.head()

City        Date  PM2.5  PM10    NO   NO2   NOx  NH3    CO    SO2  \
0  Gurugram  27/11/2015    NaN   NaN  5.28  8.37  9.57  NaN  2.88  15.92   
1  Gurugram  28/11/2015    NaN   NaN  4.71  8.00  9.74  NaN  0.47   5.78   
2  Gurugram  29/11/2015    NaN   NaN  3.10  8.04  8.69  NaN  1.67   5.26   
3  Gurugram  30/11/2015    NaN   NaN  3.90  8.21  9.54  NaN  1.45   7.51   
4  Gurugram  01/12/2015    NaN   NaN  3.36  8.13  8.86  NaN  1.45   3.72   

      O3  Benzene  Toluene  Xylene  AQI AQI_Bucket  
0   5.96      NaN      NaN     NaN  NaN        NaN  
1   8.13      NaN      NaN     NaN  NaN        NaN  
2  16.52      NaN      NaN     NaN  NaN        NaN  
3   6.29      NaN      NaN     NaN  NaN        NaN  
4  10.50      NaN      NaN     NaN  NaN        NaN

In [50]:
# Last 5 rows of dataframe
df.tail()

City        Date  PM2.5   PM10    NO    NO2   NOx    NH3    CO  \
29526  Amaravati  27/06/2020  14.50  24.43  1.53   6.53  4.72   8.97  0.55   
29527  Amaravati  28/06/2020  16.65  28.51  1.43   8.32  5.59   9.77  0.66   
29528  Amaravati  29/06/2020  20.96  32.56  1.65   9.55  6.43  14.30  0.66   
29529  Amaravati  30/06/2020  21.34  35.16  1.74  10.69  7.10  13.38  0.66   
29530  Amaravati  01/07/2020  22.00  34.00  1.50   9.68  6.40   8.45  0.59   

         SO2     O3  Benzene  Toluene  Xylene   AQI    AQI_Bucket  
29526  13.75  33.76     0.06     0.27    0.15  42.0          Good  
29527  10.86  37.34     0.10     0.43    0.12  49.0          Good  
29528  14.79  43.29     0.12     0.69    0.10  56.0  Satisfactory  
29529  14.58  45.32     0.14     1.42    0.20  61.0  Satisfactory  
29530  10.88  29.15     0.10     0.50     NaN  54.0  Satisfactory

In [51]:
# Number of Rows
df.shape[0]

29531

In [52]:
# NUmber of Columns
df.shape[1]

16

In [53]:
# Data type for each column and how many entries for each
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.6+ MB


**What can we see from this?**

We know that the data has 29,531 rows, however we can see from the code above that not all of the variables (rows) have the same amount of data entries. This means that there is missing values. We can see throughout the list the number of values varies inconsistantly, with PM10 (18391) and xylene (11422) having the lowest and CO (27472) and NO (25949) having the highest number of values. Overall, the large amount of values in each column tell us this is a reliable dataset and we can start carrying out analysis.

The data types for each entry are also correct so no formatting is needed to correct this.


#### **Preprocessing the data**



In [54]:
# Finding out the percentage of missing values per column
def missing_values_percentage_table(df):

# Finding the number of missing values per column
    missing_values = df.isnull().sum()

# Finding the percentage of missing values by dividing it by the number of data entries there should be and multiplying by 100
    missing_value_perc =  100 * missing_values / len(df)

# Concatenating the missing values and missing values %  along the top of the table
    missing_value_table = pd.concat([missing_values, missing_value_perc], axis=1)
    missing_value_table = missing_value_table.rename(columns={0: 'Missing Values', 1: '% of Total Values'})

# Sort the table by percentage of missing descending
    missing_values_table = missing_value_table.sort_values('% of Total Values', ascending=False)
    return missing_values_table

# Call the function with the dataframe 'df' and making the percentages to 2dp
missing_percentage_df = missing_values_percentage_table(df).round(2)


display(missing_percentage_df.style.background_gradient(cmap='Blues'))

**What does this table tell us?**

With Xyelene having over 50% of values missing, this column will be dropped as it cannot be reliably analysed. The columns with 10% - %40 of values missing (PM10, NH3, Toluene, Benzene, AQI / AQI_Bucket, PM2.5, NOx, O3, SO2, NO2, NO) will be interpolated to fill these quantities. CO, Date and City areat a level where they have a high level of values and good quality data analysis can be explored.

In [55]:
# Removing Xylene from the dataframe
df = df.drop('Xylene', axis=1)
df.head()

City        Date  PM2.5  PM10    NO   NO2   NOx  NH3    CO    SO2  \
0  Gurugram  27/11/2015    NaN   NaN  5.28  8.37  9.57  NaN  2.88  15.92   
1  Gurugram  28/11/2015    NaN   NaN  4.71  8.00  9.74  NaN  0.47   5.78   
2  Gurugram  29/11/2015    NaN   NaN  3.10  8.04  8.69  NaN  1.67   5.26   
3  Gurugram  30/11/2015    NaN   NaN  3.90  8.21  9.54  NaN  1.45   7.51   
4  Gurugram  01/12/2015    NaN   NaN  3.36  8.13  8.86  NaN  1.45   3.72   

      O3  Benzene  Toluene  AQI AQI_Bucket  
0   5.96      NaN      NaN  NaN        NaN  
1   8.13      NaN      NaN  NaN        NaN  
2  16.52      NaN      NaN  NaN        NaN  
3   6.29      NaN      NaN  NaN        NaN  
4  10.50      NaN      NaN  NaN        NaN